[Ready for Review] Proposal for allowing live updates to maxwell filter

crispmark opened the pull request:
We are also having issues with wanting to quickly turn on and off maxwell filters. I saw with issue #1615 that we were not the only ones having this issue. My proposal is pretty simple: expose an http endpoint `PATCH: /config` that would allow a user to supply a json containing the new filter, and update it live.

```JSON
{
  "filter":"exclude: dynamic_db.*"
}
```

I have some sample code here with it working, @osheroff please let me know if this is an avenue that you think we could pursue, and I will add tests and move the servlet outside of the metricsRegistries group.

osheroff commented:
'sok by me

crispmark commented:
@osheroff i have added some tests as well as a new config option (http_config) to support this. The endpoint does very basic validation and will return an error in the format:
```JSON
{
  "error": {
    "code": 400,
    "message": "invalid filter"
  }
}
```

both PATCH and GET endpoints are exposed on `/config` and return the serialized config (right now only filter, but can be expanded if new support for config updates is added).
```
{
  "filter": "include: *.*"
}
```
The filter is only updated if the filter string is a non-null valid filter and when the filter is updated the new filter is logged.

Let me know if this looks good or if there are any changes you want to make!

crispmark commented:
@osheroff could you take a look to see if any major contract changes are needed?

osheroff commented on src/main/java/com/zendesk/maxwell/monitoring/MaxwellConfigServlet.java:
```diff
@@ -0,0 +1,118 @@
+
+  @Override
+  public void service(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
+    if (request.getMethod().equalsIgnoreCase("PATCH")) {
```
I think we should probably support PUT/POST too.  I don't mind if the semantics aren't precisely correct, I'd rather it follow principle of least surprise.

osheroff commented:
maxwell uses tabs not spaces, one of the new files introduced had spaces

osheroff commented:
overall this looks ok.  I could see an argument for a richer representation of the filter in JSON as an array of strings (or even an array of objects, eg `[ { "exclude": { "database": "aa", "table": "bb" } }]`.  Otherwise consumers of the GET endpoint may have to implement their own filter parser.  

could also see an argument for accepting new filter configs in said JSON format *or* as strings.

Don't think that's a blocker to merging this as-is though.

osheroff commented:
if there's no filter, it should return
`{ filter: null }`
instead of what it's doing which is `{filter: ""}`

osheroff commented:
needs a scootch of documentation written.

crispmark commented:
@osheroff i have updated to make the following changes
- fixed the indentation on the servlet file
- added put and post handlers for updating the config
- added some documentation for http_config variable
- returns null for filter when filterstring is ""

osheroff commented:
ok.  merging.  thanks!
